Move the JSON error emitter to librustc_errors

Mark-Simulacrum opened the pull request:
This is done both as a cleanup (it makes little sense for this emitter to be in libsyntax), but also as part of broader work to decouple Session from librustc itself.

Along the way, this also moves SourceMap to syntax_pos, which is also nice for the above reasons, as well as allowing dropping the SourceMapper trait from code. This had the unfortunate side-effect of moving `FatalError` to rustc_data_structures (it's needed in syntax_pos, due to SourceMap, but putting it there feels somehow worse).

Mark-Simulacrum commented:
cc @Centril in case you've already been doing something similar with your breaking up libsyntax work

cc @estebank (not sure if anyone else should be pinged) for the JSON emitter changing locations -- no functional changes though

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13581&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13581/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13581/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-14T23:12:04.4579167Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-14T23:12:04.4792029Z ##[command]git config gc.auto 0
2019-11-14T23:12:04.4874555Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-14T23:12:04.4940310Z ##[command]git config --get-all http.proxy
2019-11-14T23:12:04.5078023Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66427/merge:refs/remotes/pull/66427/merge
---
2019-11-15T00:11:49.2218201Z .................................................................................................... 1500/9240
2019-11-15T00:11:55.2597594Z .................................................................................................... 1600/9240
2019-11-15T00:12:04.7443972Z .................................................................................................... 1700/9240
2019-11-15T00:12:13.4648033Z ....i............................................................................................... 1800/9240
2019-11-15T00:12:20.1526364Z ........................................................................................iiiii....... 1900/9240
2019-11-15T00:12:41.7964864Z .................................................................................................... 2100/9240
2019-11-15T00:12:44.2234312Z .................................................................................................... 2200/9240
2019-11-15T00:12:46.7241251Z .................................................................................................... 2300/9240
2019-11-15T00:12:53.7711165Z .................................................................................................... 2400/9240
---
2019-11-15T00:15:52.5691471Z .......................................................................................i............ 4700/9240
2019-11-15T00:15:59.5989407Z ...i................................................................................................ 4800/9240
2019-11-15T00:16:09.1492057Z .................................................................................................... 4900/9240
2019-11-15T00:16:14.5515363Z .................................................................................................... 5000/9240
2019-11-15T00:16:25.9232382Z ..........................................................................................ii.ii..... 5100/9240
2019-11-15T00:16:34.5735689Z .........................i.......................................................................... 5300/9240
2019-11-15T00:16:43.3327871Z .................................................................................................... 5400/9240
2019-11-15T00:16:52.1114430Z .......................................................................i............................ 5500/9240
2019-11-15T00:16:59.5701065Z .................................................................................................... 5600/9240
2019-11-15T00:16:59.5701065Z .................................................................................................... 5600/9240
2019-11-15T00:17:07.1708309Z .................................................................................................... 5700/9240
2019-11-15T00:17:16.8091854Z .........................................................ii...i..ii...........i..................... 5800/9240
2019-11-15T00:17:39.8142234Z .................................................................................................... 6000/9240
2019-11-15T00:17:47.5395953Z .................................................................................................... 6100/9240
2019-11-15T00:17:47.5395953Z .................................................................................................... 6100/9240
2019-11-15T00:17:52.5492643Z ............................................................................i..ii................... 6200/9240
2019-11-15T00:18:22.3462671Z .................................................................................................... 6400/9240
2019-11-15T00:18:24.9937798Z ............................................i....................................................... 6500/9240
2019-11-15T00:18:27.2312810Z .................................................................................................... 6600/9240
2019-11-15T00:18:29.5736166Z ..............................i..................................................................... 6700/9240
---
2019-11-15T00:23:44.4726653Z  finished in 5.764
2019-11-15T00:23:44.4930999Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:23:44.6944603Z 
2019-11-15T00:23:44.6945085Z running 156 tests
2019-11-15T00:23:47.6863601Z iiii....iii......iii..iiii...i.............................i..i..................i....i...........ii 100/156
2019-11-15T00:23:50.5457773Z .i.i..iiii..............i.........iii.i.........ii......
2019-11-15T00:23:50.5459332Z 
2019-11-15T00:23:50.5548177Z  finished in 5.133
2019-11-15T00:23:50.5554381Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:23:50.5555484Z 
---
2019-11-15T00:23:51.7711384Z  finished in 2.125
2019-11-15T00:23:51.7912521Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:23:51.9485799Z 
2019-11-15T00:23:51.9486048Z running 9 tests
2019-11-15T00:23:51.9487906Z iiiiiiiii
2019-11-15T00:23:51.9488315Z 
2019-11-15T00:23:51.9489161Z  finished in 0.158
2019-11-15T00:23:51.9681205Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:23:52.5489885Z 
---
2019-11-15T00:24:11.7684908Z  finished in 19.800
2019-11-15T00:24:11.7953336Z Check compiletest suite=debuginfo mode=debuginfo-gdb+lldb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:24:11.9953915Z 
2019-11-15T00:24:11.9954717Z running 123 tests
2019-11-15T00:24:35.6429903Z .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....ii..........iiii..........i...ii...i.......ii. 100/123
2019-11-15T00:24:40.1689373Z i.i.i......iii.i.....ii
2019-11-15T00:24:40.1690670Z 
2019-11-15T00:24:40.1692264Z  finished in 28.374
2019-11-15T00:24:40.1701134Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-15T00:24:40.1703954Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2019-11-15T00:36:46.7259746Z 
2019-11-15T00:36:46.7260464Z    Doc-tests core
2019-11-15T00:36:51.8185849Z 
2019-11-15T00:36:51.8187580Z running 2420 tests
2019-11-15T00:37:02.2580064Z ......iiiii......................................................................................... 100/2420
2019-11-15T00:37:12.4371574Z .................................................................................ii................. 200/2420
2019-11-15T00:37:36.7572319Z ...i................................................................................................ 400/2420
2019-11-15T00:37:36.7572319Z ...i................................................................................................ 400/2420
2019-11-15T00:37:46.8353453Z ...................................................i..i.................iiii........................ 500/2420
2019-11-15T00:38:05.8981192Z .................................................................................................... 700/2420
2019-11-15T00:38:15.6602437Z .................................................................................................... 800/2420
2019-11-15T00:38:25.4459758Z .................................................................................................... 900/2420
2019-11-15T00:38:35.2929994Z .................................................................................................... 1000/2420
---
2019-11-15T00:42:30.0482510Z ............................................... 300/763
2019-11-15T00:42:30.0493617Z thread '<unnamed>' panicked at 'explicit panic', src/libstd/io/stdio.rs:854:13
2019-11-15T00:42:30.1478131Z .................................................................................................... 400/763
2019-11-15T00:42:32.2324557Z .................................................................................................... 500/763
2019-11-15T00:42:32.2695171Z ....................thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.2727223Z ....thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.2734985Z thread '.<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.2762797Z ......thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.5266561Z ..........................................thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.5316915Z ...........thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:32.5328454Z .thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:1187:5
2019-11-15T00:42:34.5986703Z .......................thread '<unnamed>' panicked at 'explicit panic', src/libstd/sync/mutex.rs:629:13
2019-11-15T00:42:34.5988066Z ..thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:584:13
2019-11-15T00:42:34.5997807Z ..thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:561:13
2019-11-15T00:42:34.5998111Z thread '<unnamed>' panicked at 'explicit panic', src/libstd/sync/mutex.rs:689:13
---
2019-11-15T00:42:43.8782592Z 
2019-11-15T00:42:43.8804535Z running 1000 tests
2019-11-15T00:43:04.3304391Z i................................................................................................... 100/1000
2019-11-15T00:43:15.5069137Z .................................................................................................... 200/1000
2019-11-15T00:43:23.4258521Z ...................iii......i......i...i......i..................................................... 300/1000
2019-11-15T00:43:28.7317085Z .................................................................................................... 400/1000
2019-11-15T00:43:36.2020345Z ...........................................i..i.................................ii.................. 500/1000
2019-11-15T00:43:49.9913166Z .................................................................................................... 700/1000
2019-11-15T00:43:49.9913166Z .................................................................................................... 700/1000
2019-11-15T00:43:57.5738957Z ..........................iiii...................................................................... 800/1000
2019-11-15T00:44:12.9631696Z .................................................................................................... 900/1000
2019-11-15T00:44:20.5921482Z ................................................iiii................................................ 1000/1000
2019-11-15T00:44:20.5925830Z test result: ok. 980 passed; 0 failed; 20 ignored; 0 measured; 0 filtered out
2019-11-15T00:44:20.5926297Z 
2019-11-15T00:44:20.6020042Z  finished in 193.380
2019-11-15T00:44:20.6035184Z Testing term stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
2019-11-15T00:48:29.1543185Z    |
2019-11-15T00:48:29.1543416Z 26 | mod tests;
2019-11-15T00:48:29.1543633Z    |     ^^^^^
2019-11-15T00:48:29.1544043Z    |
2019-11-15T00:48:29.1544370Z    = help: name the file either source_map/tests.rs or source_map/tests/mod.rs inside the directory "src/libsyntax_pos"
2019-11-15T00:48:29.1549946Z error: aborting due to previous error
2019-11-15T00:48:29.1550024Z 
2019-11-15T00:48:29.1550301Z For more information about this error, try `rustc --explain E0583`.
2019-11-15T00:48:29.1650599Z error: could not compile `syntax_pos`.
---
2019-11-15T00:48:29.1755203Z   local time: Fri Nov 15 00:48:29 UTC 2019
2019-11-15T00:48:29.2370501Z   network time: Fri, 15 Nov 2019 00:48:29 GMT
2019-11-15T00:48:29.2375050Z == end clock drift check ==
2019-11-15T00:48:29.9720902Z 
2019-11-15T00:48:29.9829798Z ##[error]Bash exited with code '1'.
2019-11-15T00:48:29.9865269Z ##[section]Starting: Checkout
2019-11-15T00:48:29.9867495Z ==============================================================================
2019-11-15T00:48:29.9867590Z Task         : Get sources
2019-11-15T00:48:29.9867638Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
r? @estebank

Centril commented:
> This had the unfortunate side-effect of moving `FatalError` to rustc_data_structures (it's needed in syntax_pos, due to SourceMap, but putting it there feels somehow worse).

Hmm; it seems to me that `syntax_pos` is better than `rustc_data_structures` as the latter feels like a place solely for "in-house generic data structures which could be useful outside rustc as well". Indeed, the module docs for that crate state that the crate is for:

> Various data structures used by the Rust compiler. The intention is that code in here should be not be specific to rustc, so that it can be easily unit tested and so forth.

Centril commented on src/librustc_data_structures/lib.rs:
```diff
@@ -117,3 +117,34 @@ impl<F: Fn()> Drop for OnDrop<F> {
 #[doc(hidden)]
 pub fn __noop_fix_for_27438() {}
+
+/// Used as a return value to signify a fatal error occurred. (It is also
```
Also, it would be easier to refactor and move this around if it were contained in its own module (I know I know... micro-modules and whatnot, but I think this would be better for maintainability)

Centril commented on src/libsyntax_pos/source_map.rs:
```diff
@@ -0,0 +1,1010 @@
+//! The `SourceMap` tracks all the source code used within a single crate, mapping
```
Strange... why isn't git recognizing this as a rename of a file? Would be good not to lose blame.

Centril commented:
> This is done both as a cleanup (it makes little sense for this emitter to be in libsyntax), but also as part of broader work to decouple Session from librustc itself.

The motivation behind this change seems quite sensible. 👍

Mark-Simulacrum commented:
Yeah, I'm not a fan of the rustc_data_structures locale -- I can definitely move it to syntax_pos. I mostly decided to just give up and move it "as far back as possible" -- maybe the right strategy is to create a `rustc_utils` crate or something. I think for now I'll move it to syntax_pos though when I get a chance to edit this branch.

Mark-Simulacrum commented:
Okay I split up the move of SourceMap and git now properly detects it as a rename of the file (though the intermediate commit won't build).

Also changed the first commit to move FatalError to syntax_pos into a module (very creatively named `fatal_error`...)

Centril commented on src/librustc_errors/json/tests.rs:
```diff
@@ -40,6 +39,13 @@ impl<T: Write> Write for Shared<T> {
     }
 }
 
+fn with_default_globals(f: impl FnOnce()) {
```
Hmm; maybe this should be shared with libsyntax by moving it to libsyntax_pos?

Mark-Simulacrum replied:
Hm, my impression at a quick glance was that there are two "globals" in separate TLS variables (one declared in syntax, and one in syntax_pos). I don't think we buy anything from doing that recombination... in any case, it seems like ideally we'd cleanup the globals but that should be done separately I think.

Centril replied:
Fair :)

Centril commented:
r? @Centril

Centril commented:
Looks great; thanks! @bors r+

bors commented:
:pushpin: Commit c31a8754e3f3a9274759cb429aad4ae594d39e29 has been approved by `Centril`

<!-- @bors r=Centril c31a8754e3f3a9274759cb429aad4ae594d39e29 -->
<!-- homu: {"type":"Approved","sha":"c31a8754e3f3a9274759cb429aad4ae594d39e29","approver":"Centril"} -->
